File: code/17.letter-combinations-of-a-phone-number.php
<?php
/*
 * @lc app=leetcode id=17 lang=php
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start
class Solution {
    const LETTER_MAP = [
        '2' => ['a', 'b', 'c'],
        '3' => ['d', 'e', 'f'],
        '4' => ['g', 'h', 'i'],
        '5' => ['j', 'k', 'l'],
        '6' => ['m', 'n', 'o'],
        '7' => ['p', 'q', 'r', 's'],
        '8' => ['t', 'u', 'v',], 
        '9' => ['w', 'x', 'y', 'z'],
    ];

    /**
     * @param String $digits
     * @return String[]
     */
    function letterCombinations($digits) {
        if ($digits == '') return [];

        $hashTable = [];
        foreach (str_split($digits) as $key => $digit) {
            $list = self::LETTER_MAP[$digit];

            if ($key == 0) {
                $hashTable = $list;
                continue;
            }

            $temp = [];
            foreach($list as $letter) {
                foreach($hashTable as $letters) {
                    $temp[] = $letters. $letter;
                }
            }
            $hashTable = $temp;
        }

        return $hashTable;
    }
}
// @lc code=end

File: code/17.letter-combinations-of-a-phone-number.go
package code

/*
 * @lc app=leetcode id=17 lang=golang
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start
func letterCombinations(digits string) []string {
	if len(digits) == 0 {
		return []string{}
	}

	list := map[byte][]string{
		'2': []string{"a", "b", "c"},
		'3': []string{"d", "e", "f"},
		'4': []string{"g", "h", "i"},
		'5': []string{"j", "k", "l"},
		'6': []string{"m", "n", "o"},
		'7': []string{"p", "q", "r", "s"},
		'8': []string{"t", "u", "v"},
		'9': []string{"w", "x", "y", "z"},
	}

	// iteration
	// ans := []string{""}
	// for i := 0; i < len(digits); i++ {
	// 	cur := list[digits[i]]

	// 	new := []string{}
	// 	for _, v := range ans {
	// 		for j := 0; j < len(cur); j++ {
	// 			new = append(new, v+cur[j])
	// 		}
	// 	}

	// 	ans = new
	// }

	// return ans

	// backtracking
	ans := []string{}
	var backtracking func(combination string, next string)
	backtracking = func(combination string, next string) {
		if next == "" {
			ans = append(ans, combination)
			return
		}

		letters := list[next[0]]
		for _, letter := range letters {
			backtracking(combination+letter, next[1:])
		}
	}

	backtracking("", digits)
	return ans
}

// @lc code=end
